File: backend/app/energy.py
from datetime import datetime, timedelta
from math import ceil

WAVEPLATE_CAP = 240
WAVEPLATE_RECOVER_MINUTES = 6
WAVEPLATE_RECOVER_SECONDS = WAVEPLATE_RECOVER_MINUTES * 60
WAVEPLATE_CRYSTAL_CAP = 480
WAVEPLATE_CRYSTAL_RECOVER_MINUTES = 12
WAVEPLATE_CRYSTAL_RECOVER_SECONDS = WAVEPLATE_CRYSTAL_RECOVER_MINUTES * 60


def clamp_waveplate(value: int) -> int:
    return max(0, min(WAVEPLATE_CAP, int(value)))


def clamp_waveplate_crystal(value: int) -> int:
    return max(0, min(WAVEPLATE_CRYSTAL_CAP, int(value)))


def normalize_resources(waveplate: int, waveplate_crystal: int) -> tuple[int, int]:
    wp = max(0, int(waveplate))
    crystal = max(0, int(waveplate_crystal))
    if wp > WAVEPLATE_CAP:
        crystal += wp - WAVEPLATE_CAP
        wp = WAVEPLATE_CAP
    crystal = clamp_waveplate_crystal(crystal)
    return wp, crystal


def seconds_to_waveplate_full_from_current(waveplate: int) -> int:
    wp = clamp_waveplate(waveplate)
    if wp >= WAVEPLATE_CAP:
        return 0
    return (WAVEPLATE_CAP - wp) * WAVEPLATE_RECOVER_SECONDS


def seconds_to_waveplate_full_from_full_time(full_waveplate_at: datetime, now: datetime) -> int:
    return max(0, int((full_waveplate_at - now).total_seconds()))


def current_resources_from_full_time(
    full_waveplate_at: datetime,
    full_waveplate_crystal: int,
    now: datetime,
) -> tuple[int, int]:
    base_crystal = clamp_waveplate_crystal(full_waveplate_crystal)
    delta_seconds = int((full_waveplate_at - now).total_seconds())
    if delta_seconds > 0:
        missing = min(WAVEPLATE_CAP, ceil(delta_seconds / WAVEPLATE_RECOVER_SECONDS))
        current_wp = WAVEPLATE_CAP - missing
        return current_wp, base_crystal

    crystal_gained = max(0, int((now - full_waveplate_at).total_seconds()) // WAVEPLATE_CRYSTAL_RECOVER_SECONDS)
    return WAVEPLATE_CAP, clamp_waveplate_crystal(base_crystal + crystal_gained)


def full_time_from_current_resources(
    waveplate: int,
    current_waveplate_crystal: int,
    now: datetime,
) -> tuple[datetime, int]:
    wp, crystal = normalize_resources(waveplate, current_waveplate_crystal)
    return now + timedelta(seconds=seconds_to_waveplate_full_from_current(wp)), crystal


def full_crystal_from_current_crystal(
    current_waveplate_crystal: int,
    full_waveplate_at: datetime,
    now: datetime,
) -> int:
    current_crystal = clamp_waveplate_crystal(current_waveplate_crystal)
    if full_waveplate_at > now:
        return current_crystal
    crystal_gained = max(0, int((now - full_waveplate_at).total_seconds()) // WAVEPLATE_CRYSTAL_RECOVER_SECONDS)
    base_crystal = current_crystal - crystal_gained
    if base_crystal < 0:
        raise ValueError("current_waveplate_crystal is too small for the selected full_waveplate_at")
    return clamp_waveplate_crystal(base_crystal)


def spend_resources(
    waveplate: int,
    waveplate_crystal: int,
    cost: int,
) -> tuple[int, int]:
    wp, crystal = normalize_resources(waveplate, waveplate_crystal)
    remain = max(0, int(cost))
    use_wp = min(wp, remain)
    wp -= use_wp
    remain -= use_wp
    if remain > 0:
        crystal = max(0, crystal - remain)
    return wp, crystal


def add_resources(
    waveplate: int,
    waveplate_crystal: int,
    amount: int,
) -> tuple[int, int]:
    wp, crystal = normalize_resources(waveplate, waveplate_crystal)
    increase = max(0, int(amount))
    wp += increase
    if wp > WAVEPLATE_CAP:
        crystal = min(WAVEPLATE_CRYSTAL_CAP, crystal + (wp - WAVEPLATE_CAP))
        wp = WAVEPLATE_CAP
    return wp, crystal


def validate_full_waveplate_at(now: datetime, full_waveplate_at: datetime) -> None:
    delta_seconds = int((full_waveplate_at - now).total_seconds())
    max_seconds = seconds_to_waveplate_full_from_current(0)
    if delta_seconds > max_seconds:
        raise ValueError(f"full_waveplate_at too far in future (max {max_seconds} seconds)")


def warn_level(waveplate: int) -> str:
    wp = clamp_waveplate(waveplate)
    if wp >= 235:
        return "danger"
    if wp >= 220:
        return "warning"
    return "normal"

File: backend/internal/app/energy.go
package app

import (
	"errors"
	"fmt"
	"math"
	"time"
)

const (
	WaveplateCap            = 240
	WaveplateRecoverSeconds = 6 * 60
	WaveplateCrystalCap     = 480
	CrystalRecoverSeconds   = 12 * 60
	fourWeekDays            = 28
)

func clampWaveplate(value int) int {
	if value < 0 {
		return 0
	}
	if value > WaveplateCap {
		return WaveplateCap
	}
	return value
}

func clampWaveplateCrystal(value int) int {
	if value < 0 {
		return 0
	}
	if value > WaveplateCrystalCap {
		return WaveplateCrystalCap
	}
	return value
}

func normalizeResources(waveplate, crystal int) (int, int) {
	wp := waveplate
	if wp < 0 {
		wp = 0
	}
	cr := crystal
	if cr < 0 {
		cr = 0
	}
	if wp > WaveplateCap {
		cr += wp - WaveplateCap
		wp = WaveplateCap
	}
	return wp, clampWaveplateCrystal(cr)
}

func secondsToWaveplateFullFromCurrent(waveplate int) int {
	wp := clampWaveplate(waveplate)
	if wp >= WaveplateCap {
		return 0
	}
	return (WaveplateCap - wp) * WaveplateRecoverSeconds
}

func secondsToWaveplateFullFromFullTime(fullAt, now time.Time) int {
	seconds := int(fullAt.Sub(now).Seconds())
	if seconds < 0 {
		return 0
	}
	return seconds
}

func secondsToNextWaveplateRecoverFromFullTime(fullAt, now time.Time) int {
	delta := secondsToWaveplateFullFromFullTime(fullAt, now)
	next := delta % WaveplateRecoverSeconds
	if next == 0 {
		return WaveplateRecoverSeconds
	}
	return next
}

func currentResourcesFromFullTime(fullAt time.Time, fullCrystal int, now time.Time) (int, int) {
	baseCrystal := clampWaveplateCrystal(fullCrystal)
	deltaSeconds := int(fullAt.Sub(now).Seconds())
	if deltaSeconds > 0 {
		missing := minInt(WaveplateCap, int(math.Ceil(float64(deltaSeconds)/float64(WaveplateRecoverSeconds))))
		return WaveplateCap - missing, baseCrystal
	}

	crystalGained := maxInt(0, int(now.Sub(fullAt).Seconds())/CrystalRecoverSeconds)
	return WaveplateCap, clampWaveplateCrystal(baseCrystal + crystalGained)
}

func fullTimeFromCurrentResources(waveplate, crystal int, now time.Time) (time.Time, int) {
	wp, cr := normalizeResources(waveplate, crystal)
	return now.Add(time.Duration(secondsToWaveplateFullFromCurrent(wp)) * time.Second), cr
}

func fullTimeFromCurrentWaveplateAndNextRecover(waveplate, nextRecoverSeconds int, now time.Time) time.Time {
	wp := clampWaveplate(waveplate)
	if wp >= WaveplateCap {
		return now
	}
	missing := WaveplateCap - wp
	next := maxInt(1, minInt(WaveplateRecoverSeconds, nextRecoverSeconds))
	totalSeconds := ((missing - 1) * WaveplateRecoverSeconds) + next
	return now.Add(time.Duration(totalSeconds) * time.Second)
}

func fullCrystalFromCurrentCrystal(currentCrystal int, fullAt, now time.Time) (int, error) {
	cr := clampWaveplateCrystal(currentCrystal)
	if fullAt.After(now) {
		return cr, nil
	}
	crystalGained := maxInt(0, int(now.Sub(fullAt).Seconds())/CrystalRecoverSeconds)
	base := cr - crystalGained
	if base < 0 {
		return 0, errors.New("current_waveplate_crystal is too small for the selected full_waveplate_at")
	}
	return clampWaveplateCrystal(base), nil
}

func spendResources(waveplate, crystal, cost int) (int, int) {
	wp, cr := normalizeResources(waveplate, crystal)
	remain := maxInt(0, cost)
	useWP := minInt(wp, remain)
	wp -= useWP
	remain -= useWP
	if remain > 0 {
		cr = maxInt(0, cr-remain)
	}
	return wp, cr
}

func addResources(waveplate, crystal, amount int) (int, int) {
	wp, cr := normalizeResources(waveplate, crystal)
	increase := maxInt(0, amount)
	wp += increase
	if wp > WaveplateCap {
		cr = minInt(WaveplateCrystalCap, cr+(wp-WaveplateCap))
		wp = WaveplateCap
	}
	return wp, cr
}

func validateFullWaveplateAt(now, fullAt time.Time) error {
	deltaSeconds := int(fullAt.Sub(now).Seconds())
	maxSeconds := secondsToWaveplateFullFromCurrent(0)
	if deltaSeconds > maxSeconds {
		return fmt.Errorf("full_waveplate_at too far in future (max %d seconds)", maxSeconds)
	}
	return nil
}

func warnLevel(waveplate int) string {
	wp := clampWaveplate(waveplate)
	if wp >= 235 {
		return "danger"
	}
	if wp >= 220 {
		return "warning"
	}
	return "normal"
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}
